### PR TITLE
Fix for some compiler warnings in parquet/page_decode.cuh

### PR DESCRIPTION
**Pull request by etseidl**
## Description
Clangd generates several warnings/errors in cpp/src/io/parquet/page_decode.cuh. One is in regards to a lambda argument shadowing a captured value. The others involve the use of `thrust::optional::value()` in device code...unlike pretty much every other member function, `value()` lacks the `__device__` decorator. This PR replaces two usages of `value()` with `operator*()` which does have the `__device__` decorator.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by davidwendt — APPROVED** [review]
Actually, `operator()*` is preferred in device code since `value()` may throw an exception.

**Review by ttnghia — APPROVED** [review]

**Review by pmattione-nvidia — APPROVED** [review]

File: cpp/src/io/parquet/page_decode.cuh [hunk 1]
```diff
@@ -924,7 +924,7 @@ inline __device__ uint32_t InitLevelSection(page_state_s* s,
 
   auto start = cur;
 
-  auto init_rle = [s, lvl, end, level_bits](uint8_t const* cur, uint8_t const* end) {
+  auto init_rle = [s, lvl, level_bits](uint8_t const* cur, uint8_t const* end) {
```
Comment (vuule) [inline cpp/src/io/parquet/page_decode.cuh:927]: The behavior when a captured variable and a parameter have the same name is interesting. Thanks to godbolt, I found that all options are possible, with different compilers -  captured variable value is used, or parameter value is used, or compilation fails. Fun!

**Review by vuule — APPROVED** [review]

File: cpp/src/io/parquet/page_decode.cuh (hunk 1 above)
Comment (ttnghia) [reply cpp/src/io/parquet/page_decode.cuh:927]: Wow that's interesting. We were in the same situation before, and the code runs differently on different architectures (due to compiled into different archs?).

Comment (etseidl) [reply cpp/src/io/parquet/page_decode.cuh:927]: Thank you clangd! :smile:

Comment (vuule) [reply cpp/src/io/parquet/page_decode.cuh:927]: I think nvcc stayed consistent in prioritizing the parameter over the capture. But I found online that other compilers disagree.

Comment (ttnghia) [reply cpp/src/io/parquet/page_decode.cuh:927]: We found that the capture overshadows the previous variable on H100+V100 while on other GPUs it doesn't.

**Pull request merged**